shared/data_split.py

```diff
@@ -42,6 +53,19 @@ def get_dirichlet_split_indexes(labels: list[int], num_split, alpha):
             my_index_end = math.floor(end_prob * num_label_indices)
             split_label_idx[rank].extend(label_indices[my_index_start:my_index_end])
 
+    if balance_approach:
+        expected_length = len(labels) // num_split
+        diff_per_rank = [len(v) - expected_length for v in split_label_idx]
+        extra_labels = []
+        for rank in np.argsort(diff_per_rank)[::-1]:  # from the client has most extra labels
+            if diff_per_rank[rank] > 0:
+                np.random.shuffle(split_label_idx[rank])
+                extra_labels.extend(split_label_idx[rank][-diff_per_rank[rank] :])
+                split_label_idx[rank] = split_label_idx[rank][: -diff_per_rank[rank]]
+            elif diff_per_rank[rank] < 0:
+                split_label_idx[rank].extend(extra_labels[: abs(diff_per_rank[rank])])
+                extra_labels = extra_labels[abs(diff_per_rank[rank]) :]
+
     return [sorted(v) for v in split_label_idx]
```

Comment:
Wondering if we need to shuffle `extra_labels` right before we try to extend the groups whose elements are less than expected length?

current 'extra_labels' looks like this: 
![image](https://github.com/user-attachments/assets/ee1aaec9-a043-48b2-9c73-dfd3c145b420)

If not shuffling before try to append, highly likely most elements to be appended will come from just 1 group before the balancing.

Reply:
ok. SG. shuffled.